Streamline optimized baseline guide (#1249)
llm-d/llm-d: guides/optimized-baseline/README.md

Reviewer: ```suggestion
The optimized-baseline defaults to two main routing criteria:
- **Prefix-cache aware** using the [prefix cache scorer](https://github.com/llm-d/llm-d-inference-scheduler/tree/main/pkg/epp/framework/plugins/scheduling/scorer/prefix), which scores candidate endpoints by estimating prompt prefix cache reuse on each model server. 
- *Load-aware* using both the [kv-cache utilization](https://github.com/llm-d/llm-d-inference-scheduler/tree/main/pkg/epp/framework/plugins/scheduling/scorer/kvcacheutilization) and the [queue size](https://github.com/llm-d/llm-d-inference-scheduler/tree/main/pkg/epp/framework/plugins/scheduling/scorer/queuedepth) scorers.

```

## Overview

This guide deploys the recommended out of the box [scheduling configuration](https://github.com/llm-d/llm-d-inference-scheduler/blob/main/docs/architecture.md) for most vLLM and SGLang deployments, reducing tail latency and increasing throughput through load-aware and prefix-cache aware balancing.

The optimized-baseline defaults to two main routing criteria:

- **Prefix-cache aware** using the [prefix cache scorer](https://github.com/llm-d/llm-d-inference-scheduler/tree/main/pkg/epp/framework/plugins/scheduling/scorer/prefix), which scores candidate endpoints by estimating prompt prefix cache reuse on each model server. 

> - *Load-aware* using both the [kv-cache utilization](https://github.com/llm-d/llm-d-inference-scheduler/tree/main/pkg/epp/framework/plugins/scheduling/scorer/kvcacheutilization) and the [queue size](https://github.com/llm-d/llm-d-inference-scheduler/tree/main/pkg/epp/framework/plugins/scheduling/scorer/queuedepth) scorers.

Reviewer: ```suggestion
- **Load-aware** using both the [kv-cache utilization](https://github.com/llm-d/llm-d-inference-scheduler/tree/main/pkg/epp/framework/plugins/scheduling/scorer/kvcacheutilization) and the [queue size](https://github.com/llm-d/llm-d-inference-scheduler/tree/main/pkg/epp/framework/plugins/scheduling/scorer/queuedepth) scorers.
```

| CPU                  | `modelserver/cpu/vllm/`    | Intel/AMD, 64 cores + 64GB RAM per replica |

> [!NOTE]
> Some hardware variants use reduced configurations (fewer replicas, smaller models) to enable CI testing for compatibility and regression checks. These configurations are maintained by their respective hardware vendors and are not guaranteed as production-ready examples. Users deploying on non-default hardware should review and adjust the configurations for their environment.

## Prerequisites

- Have the [proper client tools installed on your local system](../../helpers/client-setup/README.md) to use this guide.
- Checkout llm-d repo:

  ```bash
    export branch="main" # branch, tag, or commit hash
    git clone https://github.com/llm-d/llm-d.git && cd llm-d && git checkout ${branch}
  ```

- (Optional) Have the [Monitoring stack](../../docs/monitoring/README.md) installed on your system.

## Installation Instructions

### 1. Prepare a Target Namespace

- Create a target namespace for the installation.

  ```bash
  export NAMESPACE=llm-d-optimized-baseline
  kubectl create namespace ${NAMESPACE}
  ```

### 2. Deploy the Standalone Inference Scheduler

> This deploys the iinference scheduler with an Envoy sidecar. For gateway deployments, see [Gateway recipes](../recipes/gateway).

Reviewer: ```suggestion
This deploys the inference scheduler with an Envoy sidecar. For gateway deployments, see the next collapsed section.
```

```bash
helm install optimized-baseline-scheduler \
  oci://registry.k8s.io/gateway-api-inference-extension/charts/standalone \
  -f guides/recipes/scheduler/base.values.yaml \
  -f guides/optimized-baseline/scheduler/optimized-baseline.values.yaml \
  -n ${NAMESPACE} --version v1.4.0
```

> 

Reviewer: https://github.com/llm-d/llm-d/pull/1259 consolidates the gateway guides into one that should compose with the well-lit paths guides. Here we will add the following collapsable section (will do it in #1259):

<details>
<summary>#### Gateway Mode</summary>

If you would like to employ a Kubernetes Gateway managed proxy instead of a standalone one, then don't apply the standalone helm chart, and do the following:

1. Deploy a Gateway, see [the gateway guides](../prereq/gateways) for manifests and step by step deployment. 
2. Deploy the Inference Scheduler helm chart as follows:

```bash
helm install optimized-baseline \
  oci://registry.k8s.io/gateway-api-inference-extension/charts/inferencepool  \
  -f guides/recipes/scheduler/base.values.yaml \
  -f guides/optimized-baseline/scheduler/optimized-baseline.values.yaml \
  --set provider.name=[none|gke|istio] \
  --set experimentalHttpRoute.enabled=true \
  --set experimentalHttpRoute.inferenceGatewayName=llm-d-inference-gateway \
  -n ${NAMESPACE} --version v1.4.0
```
</details>

### 3. Deploy the Model Server

Apply the Kustomize overlays for your specific backend (defaulting to NVIDIA GPU / vLLM):

```bash
kubectl apply -n ${NAMESPACE} -k guides/optimized-baseline/modelserver/gpu/vllm/
```

> [TEMPORARY] For GKE, run this instead:

Reviewer: Until when?

```bash

> kubectl apply -n ${NAMESPACE} -k guides/optimized-baseline/modelserver/gpu/vllm-gke/

Reviewer: I am getting the following error when applying this:

```
error: resource mapping not found for name: "optimized-baseline-nvidia-gpu-vllm-decode" namespace: "" from "guides/optimized-baseline/modelserver/gpu/vllm/": no matches for kind "PodMonitor" in version "monitoring.coreos.com/v1"
ensure CRDs are installed first
```

Author: Updated steps, monitoring is an optional step now.

> # Optimized Baseline

Reviewer: Please update `scheduler/optimized-baseline.values.yaml`, the flags `pluginsConfigFile` and `pluginsCustomConfig` must be under `inferenceExtension` not under `inferencePool`.

Also, please set verbosity to 2 (our actual Default) in the base under recipes

Reviewer: `--set inferenceExtension.flags.v=2`

> [!NOTE]
> Some hardware variants use reduced configurations (fewer replicas, smaller models) to enable CI testing for compatibility and regression checks. These configurations are maintained by their respective hardware vendors and are not guaranteed as production-ready examples. Users deploying on non-default hardware should review and adjust the configurations for their environment.

## Prerequisites

- Have the [proper client tools installed on your local system](../../helpers/client-setup/README.md) to use this guide.
- Checkout llm-d repo:

  ```bash
    export branch="main" # branch, tag, or commit hash
    git clone https://github.com/llm-d/llm-d.git && cd llm-d && git checkout ${branch}
  ```

## Installation Instructions

### 1. Prepare a Target Namespace

- Create a target namespace for the installation.

  ```bash
  export NAMESPACE=llm-d-optimized-baseline
  kubectl create namespace ${NAMESPACE}
  ```

### 2. Deploy the Standalone Inference Scheduler

This deploys the inference scheduler with an Envoy sidecar. For gateway deployments, see the next collapsed section.

```bash

> helm install optimized-baseline-scheduler \

Reviewer: ```suggestion
helm install optimized-baseline \
```

Reviewer: The epp deployment will automatically have the suffix `-epp`

## Default Configuration

| Parameter          | Value                                                   |
| ------------------ | ------------------------------------------------------- |
| Model              | [Qwen/Qwen3-32B](https://huggingface.co/Qwen/Qwen3-32B) |
| Replicas           | 8                                                       |
| Tensor Parallelism | 2                                                       |
| GPUs per replica   | 2                                                       |
| Total GPUs         | 16                                                      |

### Supported Hardware Backends

This guide includes configurations for the following accelerators:

| Backend              | Directory                  | Notes                                      |
| -------------------- | -------------------------- | ------------------------------------------ |
| NVIDIA CUDA          | `modelserver/cuda/vllm/`   | Default configuration                      |

> | NVIDIA CUDA (SGLang) | `modelserver/cuda/sglang/` | SGLang inference server                    |

Reviewer: Calling this `Nvidia CUDA` is not aligned with the other backends which uses the accelerator name; whether we like it or not, `gpu` as a term is commonly understood as nvidia gpus, so we should just use `NVIDIA GPU` and the kustomize path to be `modelserver/gpu/vllm|sglang` instead of cuda; for AMD gpus we call it `amd`

| AMD ROCm             | `modelserver/amd/vllm/`    | AMD GPU                                    |
| Intel XPU            | `modelserver/xpu/vllm/`    | Intel Data Center GPU Max 1550+            |
| Intel Gaudi (HPU)    | `modelserver/hpu/vllm/`    | Gaudi 1/2/3 with DRA support               |
| Google TPU v6e       | `modelserver/tpu-v6/vllm/` | GKE TPU                                    |
| Google TPU v7        | `modelserver/tpu-v7/vllm/` | GKE TPU                                    |
| CPU                  | `modelserver/cpu/vllm/`    | Intel/AMD, 64 cores + 64GB RAM per replica |

> > **Note on hardware variant configurations:** Some hardware variants use reduced configurations (fewer replicas, smaller models) to enable CI testing for compatibility and regression checks. These configurations are maintained by their respective hardware vendors and are not guaranteed as production-ready examples. Users deploying on non-default hardware should review and adjust the configurations for their environment.

Reviewer: ```suggestion
> [!NOTE]
> Some hardware variants use reduced configurations (fewer replicas, smaller models) to enable CI testing for compatibility and regression checks. These configurations are maintained by their respective hardware vendors and are not guaranteed as production-ready examples. Users deploying on non-default hardware should review and adjust the configurations for their environment.
```

    git clone https://github.com/llm-d/llm-d.git && cd llm-d && git checkout ${branch}
  ```

- (Optional) Have the [Monitoring stack](../../docs/monitoring/README.md) installed on your system.

## Installation Instructions

### 1. Prepare a Target Namespace

- Create a target namespace for the installation.

  ```bash
  export NAMESPACE=llm-d-optimized-baseline
  kubectl create namespace ${NAMESPACE}
  ```

### 2. Deploy the Standalone Inference Scheduler

This deploys the EPP scheduler with an Envoy sidecar. For gateway deployments, see [Gateway recipes](../recipes/gateway).

```bash
helm install optimized-baseline-scheduler \
  oci://registry.k8s.io/gateway-api-inference-extension/charts/standalone \
  -f guides/recipes/scheduler/base.values.yaml \
  -f guides/optimized-baseline/scheduler/optimized-baseline.values.yaml \
  -n ${NAMESPACE} --version v1.4.0
```

### 3. Deploy the Model Server

Build and apply the Kustomize overlays for your specific backend (defaulting to NVIDIA CUDA / vLLM):

```bash
kubectl apply -n ${NAMESPACE} -k guides/optimized-baseline/modelserver/nvidia-gpu/vllm/
```

## Verification

### 1. Port-Forward to the Scheduler Service

> Expose the standalone EPP service to your local environment:

Reviewer: ```suggestion
Expose the standalone inference scheduler service to your local environment:
```

> **Note on hardware variant configurations:** Some hardware variants use reduced configurations (fewer replicas, smaller models) to enable CI testing for compatibility and regression checks. These configurations are maintained by their respective hardware vendors and are not guaranteed as production-ready examples. Users deploying on non-default hardware should review and adjust the configurations for their environment.

## Prerequisites
- Have the [proper client tools installed on your local system](../../helpers/client-setup/README.md) to use this guide.
- (Optional) Have the [Monitoring stack](../../docs/monitoring/README.md) installed on your system.

## Installation Instructions

> ### 1. Prepare a Target Namespace

Reviewer 2: I actually don’t think this namespace is needed

Author: It's not strictly required, but defaulting to "default" is a bit dangerous (collision with other well-lit path guides for example).

Reviewer 3: +1 what @liu-cong says here, its not like helmfile where we could provide a reasonable default

- Create a target namespace for the installation.

  ```bash
  export NAMESPACE=llm-d-optimized-baseline
  kubectl create namespace ${NAMESPACE}
  ```

> - [Create the `llm-d-hf-token` secret in your target namespace with the key `HF_TOKEN` matching a valid HuggingFace token](../../helpers/hf-token.md) to pull models.

Reviewer 2: HF token is not needed either

> ### 2. Deploy the Standalone Inference Scheduler

Reviewer 2: Deploy the “Inference Router” I think is the new term

Author: I think we decided to keep it until we do the repo renaming, no?

> This deploys the EPP scheduler with an Envoy sidecar. For advanced gateway deployments, see [Gateway recipes](../recipes/gateway).

Reviewer 2: I would avoid using the term “advanced”

```bash
helm install optimized-baseline-scheduler \

>   oci://registry.k8s.io/gateway-api-inference-extension/charts/standalone \

Reviewer 2: These paths are now not right relative to the current directory

Reviewer 3: Agreed we should probably do some git processing to extract the repo root, and then structure the value paths off that

Author: Added git commands, and all guides should be executed from the root folder to simply relative links.

  -f guides/recipes/scheduler/base.values.yaml \
  -f guides/recipes/scheduler/features/monitoring.values.yaml \
  -f guides/optimized-baseline/scheduler/optimized-baseline.values.yaml \
  -n ${NAMESPACE} --version v1.4.0
```

### 3. Deploy the Model Server
Build and apply the Kustomize overlays for your specific backend (defaulting to NVIDIA CUDA / vLLM):
```bash
kustomize build guides/optimized-baseline/modelserver/nvidia-gpu/vllm/ | kubectl apply -n ${NAMESPACE} -f -
```

> ## Verification

Reviewer 3: I would suggest that we move all guide verification to a single README doc that walks trhough verifying ALL guides because the process is roughly the same for all of them

Author: Verification is a simple curl command, it's Ok to duplicate the commands in each guide. Directing to another file is very distracting.

Reviewer 2: Agree with cong

### 1. Port-Forward to the Scheduler Service
Expose the standalone EPP service to your local environment:
```bash
kubectl port-forward -n ${NAMESPACE} svc/optimized-baseline-scheduler 8000:9002
```

### 2. Send Test Requests
In a separate terminal, verify model availability and inference:

**List Available Models:**
```bash
curl -s http://localhost:8000/v1/models | jq
```

**Send a Completion Request:**
```bash
curl -X POST http://localhost:8000/v1/completions \
  -H 'Content-Type: application/json' \
  -d '{
        "model": "Qwen/Qwen3-32B",
        "prompt": "How are you today?"
      }' | jq
```

## Benchmarking

> To evaluate the profile performance under load:

Reviewer 2: It’s probably worthwhile to give a brief explanation of what this benchmark is (ie it’s a multi-turn benchmark suite)

Reviewer 3: We should link to the benchmarking doc for this

kubectl apply -n ${NAMESPACE} -k guides/optimized-baseline/modelserver/gpu/vllm/
```

> [TEMPORARY] For GKE, run this instead:

Reviewer: You still have the gke specific vllm manifests, I hope we don't actually need them, but if we do, why are we removing this.

Author: The kustomize patch folder names are very discoverable (e.g., `vllm-gke`) so I don't want to document all the patch paths. We document base vllm on gpu as default, and let the users discover the appropriate one for their environment.
